### PR TITLE
Add duration info to pipeline steps (tree view) UI

### PR DESCRIPTION
**Pull request by svanoort**
Adds timing info to tree view to help figure out what is taking time in a pipeline.  Spinoff from Jenkins world talks. 

- [x] Gets basic timing info for nodes
- [x] For blocks, they use the end node to compute total duration (yay!)
- [x] Handle cases where no start time is present
- [x] Provide a clean string formatting for elapsed time
- [x] Show if time is in the block or just the node self (cue for users on hoo to interpret)
- [x] Bugfix: figure out why last node or pair of nodes in a block is sometimes not getting timing info?

Manually tested with different (often complex) nesting structures, and with parallels.  Bug above is only one seen.  If no TimingAction is present on one of the flownodes (possible if there's a failure with loading actions), then no timing is shown for that one, rather than invalid info.

### CODE REVIEW

**Comment by jglick** [comment]
Sounds useful.

**Comment by svanoort** [comment]
**Screenshot in action:** 

screen shot 2016-09-09 at 4 47 27 pm

@reviewbybees

**Comment by jglick** [comment]
> Show if time is in the block or just the node self

What about a step that takes some significant time in itself *and* in one or more blocks?

**Comment by jglick** [comment]
Did you test with steps which run multiple blocks, such as {{waitUntil}} or {{retry}}?

**Comment by jglick** [comment]
```
Nullcheck of newRow at line 217 of value previously dereferenced in org.jenkinsci.plugins.workflow.support.visualization.table.FlowGraphTable.order(FlowGraphTable$Row) [org.jenkinsci.plugins.workflow.support.visualization.table.FlowGraphTable, org.jenkinsci.plugins.workflow.support.visualization.table.FlowGraphTable] At FlowGraphTable.java:[line 217]Redundant null check at FlowGraphTable.java:[line 222] RCN_REDUNDANT_NULLCHECK_WOULD_HAVE_BEEN_A_NPE
```

**Comment by svanoort** [comment]
> What about a step that takes some significant time in itself and in one or more blocks?

@jglick Could be a good future enhancement.  We should talk about use cases to see if it is worth displaying specially (and how).  Adds some significant complexity though, so would want to do it with good reason only. 

> multiple blocks

Why would that be an issue, if it only looks at BlockStartNode & BlockEndNode?

> FindBugs

Already fixed.

**Comment by jglick** [comment]
> Why would that be an issue

I have no idea how your implementation works, I am just asking whether you did any tests of this possibly interesting case—if so, your screenshot does not show it.

**Comment by jglick** [comment]
:bee:

**Comment by svanoort** [comment]
> I am just asking whether you did any tests of this possibly interesting case—if so, your screenshot does not show it.

I have not, but would expect the implementation to handle them seamlessly, and it turns out it does!  *grin*

screen shot 2016-09-09 at 5 07 46 pm

File: src/main/java/org/jenkinsci/plugins/workflow/support/visualization/table/FlowGraphTable.java [hunk 1]
```diff
@@ -67,13 +71,13 @@ public void build() {
      */
     private Map<FlowNode, Row> createAllRows() {
         heads = execution.getCurrentHeads();
-        FlowGraphWalker walker = new FlowGraphWalker();
-        walker.addHeads(heads);
+        DepthFirstScanner scanner = new DepthFirstScanner();
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/workflow/support/visualization/table/FlowGraphTable.java]: use final, Luke

**Comment by abayer** [comment]
works for me :bee: - sorry I just had @jglick do a release of `workflow-support` like 30 minutes ago. =)

File: src/main/java/org/jenkinsci/plugins/workflow/support/visualization/table/FlowGraphTable.java [hunk 2]
```diff
@@ -243,7 +279,27 @@ public String getDisplayName() {
...
+            return this.startTimeMillis;
+        }
+
+        public long getDurationMillis() {
+            return this.durationMillis;
+        }
+
+        public String getDurationString() {
+            if (this.durationMillis == 0) {
+                return "";
```
Comment (oleg-nenashev) [inline src/main/java/org/jenkinsci/plugins/workflow/support/visualization/table/FlowGraphTable.java]: Hmm, maybe "N/A"?

Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/workflow/support/visualization/table/FlowGraphTable.java]: It's hidden anyway (not shown if 0), so kind of irrelevant

**Comment by oleg-nenashev** [comment]
🐝 anyway

File: src/main/java/org/jenkinsci/plugins/workflow/support/visualization/table/FlowGraphTable.java [hunk 3]
```diff
@@ -67,13 +69,13 @@ public void build() {
      */
     private Map<FlowNode, Row> createAllRows() {
         heads = execution.getCurrentHeads();
-        FlowGraphWalker walker = new FlowGraphWalker();
-        walker.addHeads(heads);
+        final DepthFirstScanner scanner = new DepthFirstScanner();
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/support/visualization/table/FlowGraphTable.java]: Does not work; cf. #13.

**Pull request merged**